crates/rolldown_common/src/types/symbol_ref.rs

```diff
@@ -43,37 +42,26 @@ impl SymbolRef {
     }
   }
 
-  pub fn is_created_by_import_from_external(&self, modules: &IndexModules) -> bool {
-    self.inner_is_created_by_import_from_external(modules, &mut FxHashSet::default())
-  }
-  fn inner_is_created_by_import_from_external(
-    self,
+  pub fn is_created_by_import_stmt_that_target_external(
+    &self,
+    db: &SymbolRefDb,
     modules: &IndexModules,
-    visited: &mut FxHashSet<SymbolRef>,
   ) -> bool {
-    let is_not_inserted_before = visited.insert(self);
-    if !is_not_inserted_before {
-      // We are in a cycle
-      return false;
-    }
+    let canonical_ref = db.par_canonical_ref_for(*self);
 
-    let Module::Normal(owner) = &modules[self.owner] else { return false };
+    let Module::Normal(owner) = &modules[canonical_ref.owner] else { return false };
 
-    let Some(named_import) = owner.named_imports.get(&self) else {
+    let Some(named_import) = owner.named_imports.get(self) else {
       return false;
     };
 
     let rec = &owner.import_records[named_import.record_id];
 
     match &modules[rec.resolved_module] {
-      Module::Normal(normal) => {
-        let Specifier::Literal(imported) = &named_import.imported else {
-          return false;
-        };
-        let Some(named_export) = normal.named_exports.get(imported) else {
-          return false;
-        };
-        named_export.referenced.is_created_by_import_from_external(modules)
+      Module::Normal(_) => {
```

Comment:
If this branch is never reachable, we should use `unreachable!` to holds you assumption.

Reply:
Yeah. Don't panic does cause invalid state, in this case, the invalid state would cause generate non-short export pattern for external imported symbols. Considering the trade-off, I didn't throw panic.

However, I do prefer throwing panic now. We should stop this invalid state as soon as possible, because there might be something wrong already.